google-cloud-firestore/src/main/java/com/google/cloud/firestore/UpdateBuilder.java

```diff
@@ -675,4 +712,67 @@ boolean isEmpty() {
   public int getMutationsSize() {
     return writes.size();
   }
+
+  BatchState getState() {
+    return state;
+  }
+
+  HashSet<String> getDocPaths() {
+    return docPaths;
+  }
+
+  int getOpCount() {
+    return resultsMap.size();
+  }
+
+  private ApiFuture<WriteResult> processOperation(DocumentReference documentReference) {
+    Preconditions.checkState(
+        allowDuplicateDocs() || !docPaths.contains(documentReference.getPath()),
+        "Batch should not contain writes to the same document");
+    Preconditions.checkState(state == BatchState.OPEN, "Batch should be OPEN when adding writes");
+    docPaths.add(documentReference.getPath());
+    SettableApiFuture<WriteResult> result = SettableApiFuture.create();
+    resultsMap.put(getOpCount(), result);
+
+    if (getOpCount() == maxBatchSize) {
+      state = BatchState.READY_TO_SEND;
+    }
+
+    return result;
+  }
+  /** Resolves the individual operations in the batch with the results. */
+  void processResults(List<BatchWriteResult> results, @Nullable Exception error) {
+    if (error == null) {
+      for (int i = 0; i < getOpCount(); ++i) {
```

Comment:
You are looping over the resultsMap, but this is not apparent. This would be easier to read if you inlined `getOpCount` here and below (and maybe we can remove it altogether?).